### PR TITLE
refactor(gossip) Add methods, data structures, and edits to improve memory safety in gossip 

### PR DESCRIPTION
**Pull request by yewman**
**Memory Leaks**

- Found a few more memory leaks while thinking about clone / deinit implementation. 
- Implemented basic fix - call bincode.free 
- Ideally should add test cases for unhappy paths

**ThreadSafeContactInfo**
- In a number of places, ContactInfo was being passed around outside of lock boundaries for purposes not requiring access to the heap memory contained within. Whilst these usages were not causing any errors, creating and using an explicitly thread safe type seems like a good idea as makes the code more robust, simpler to understand, and easier to debug. 

**Clone**
- Implement clone methods for instances where we need access to the heap allocated memory of gossip values outside of a table lock scope.

**Other**
- Add saturating subs for timestamp operations
- Remove parameter to processMessages - no need to parse a parameter which is contained in self

### CODE REVIEW

File: src/gossip/ping_pong.zig
Comment (InKryption) [inline src/gossip/ping_pong.zig:138]: In contrast but not opposition to my other comment, I think this kind of thing instead warrants extracting the type into a named local declaration somewhere to avoid this kind of duplication, ie:
```zig
const Pings = LruCache(.non_locking, PubkeyAndSocketAddr, std.time.Instant);
const Pongs = LruCache(.non_locking, PubkeyAndSocketAddr, std.time.Instant);
```
maybe only one type like `PingsOrPongs` if the type of these fields are "intertwined" (as in, probably never going to differ).

**Review by InKryption — COMMENTED** [review]
Just a few notes, otherwise LGTM.

**Comment by InKryption** [comment]
Th failing CI is due to #157.

File: src/gossip/ping_pong.zig
Comment (yewman) [reply src/gossip/ping_pong.zig:118]: I generally prefer it as well, I noticed that the rest of the gossip lib tends to use the fully-qualified approach so I thought I'd update it for consistency.

Comment (InKryption) [reply src/gossip/ping_pong.zig:118]: There is some inconsistency, some code fully-qualifies, other code does stuff like `std.ArrayList` and similar.

File: src/gossip/service.zig
Comment (0xNineteen) [inline src/gossip/service.zig:1214]: should keep this as a pointer in this for loop (`*pull_request`) - without the pointer it will likely lead to a stack copy which we dont need since we only access the fields

**Review by 0xNineteen — CHANGES_REQUESTED** [review]

File: src/gossip/service.zig
Comment (yewman) [reply src/gossip/service.zig:910]: Yeah I left this one because I actually wanted to briefly discuss the best approach here. We can either move the logic responsible for building messages into the scope of the read lock, or clone the entries we pull out of the gossip table. This is an instance where we cant use thread safe contact info because we are serialising the heap allocated data when constructing the messages. I don't have a strong intuition over which approach would be better performance wise, so unless one of the two above options jumps out at you as the clear choice perhaps Ill need to figure out a way to benchmark

File: src/gossip/service.zig
Comment (yewman) [reply src/gossip/service.zig:771]: Quick question, there is not default implementation for starting a std.time.Timer at an effective 'zero' state, thus timing variables which are referenced before their non default assignment won't fit the Timer well. I think it makes sense to just create a function to create this 'zero' state for the std.Timer but thought I'd check before actually implementing it

**Review by 0xNineteen — COMMENTED** [review]
lgtm - should be good to merge after you add the docs

**Review by 0xNineteen — APPROVED** [review]
lgtm - should get @dnut approval too since he has a lot of context around this problem

File: src/gossip/table.zig
Comment (dnut) [inline src/gossip/table.zig:413]: In all of the usages where null is passed here, the return contains a pointer to data owned by the table.

- getVotesWithCursor
- getEpochSlotsWithCursor
- getDuplicateShredsWithCursor

So it would be unsafe to use the data returned by those methods across a lock boundary.

Those methods are not actually used anywhere. My first instinct is to remove them. Not sure why we have unused methods. I'm guessing it's for feature parity with the agave implementation. In that case, I would probably just change them to accept an allocator. It seems preferable to default to safety and require close inspection to be fast, rather than defaulting to speed and requiring close inspection to be safe.

Also I would probably make this method private and document that it may be unsafe to use the return value across lock boundaries if the allocator is not set.

File: src/gossip/service.zig
Comment (dnut) [inline src/gossip/service.zig:760]: Is there a disadvantage to using `try` here? Returning an error seems preferable to `unreachable` when performance is not a concern.

File: src/gossip/service.zig
Comment (dnut) [inline src/gossip/service.zig:837]: We could probably start replacing all of this stuff with `Duration` at some point. Doesn't need to be in this PR, just thinking out loud.

File: src/gossip/service.zig
Comment (dnut) [inline src/gossip/service.zig:1228]: nice

**Review by dnut — COMMENTED** [review]
Looks good, I just have a few minor concerns.

I like the approach of just returning a lean contact info with the bare minimum data, instead of requiring more allocations. It actually moves us back away from the new ContactInfo, closer towards something resembling LegacyContactInfo. If we go all-in on this approach, at some point we could probably get rid of the contact info mapping that I added to facilitate the migration to ContactInfo.
```
converted_contact_infos: AutoArrayHashMap(Pubkey, ContactInfo),
```
I can imagine some approaches where a caller could specify exactly the minimum amount of data they need about a contact. Something that feels like graphql.

File: src/bincode/bincode.zig
Comment (InKryption) [inline src/bincode/bincode.zig:160]: This doesn't actually respect the added field in `FieldConfig`, making either one or the other redundant, or the former is missing code that actually uses it.

But beyond that, both this and the added field in `FieldConfig` is what I was wanting to avoid. It's this kind of functionality that should, in my view, be added to a field via a custom `serialize` function, leaving bincode logic simple.

**Review by InKryption — COMMENTED** [review]
Need to go over some details concerning bincode.

File: src/bincode/bincode.zig
Comment (0xNineteen) [reply src/bincode/bincode.zig:160]: for more context - the approach ink's talking about is similar to [this](https://github.com/Syndica/sig/blob/20a6d3dd4a2f47ceea50ba2faad9e622a60ecc54/src/bincode/arraylist.zig#L17) - where after the serialization, you populate the cache in the same `serialize` function

Comment (InKryption) [reply src/bincode/bincode.zig]: We spoke about this in a huddle, we'll proceed with this as is, but I have a plan to improve the situation, which is out of scope for this PR.

File: src/gossip/service.zig
Comment (0xNineteen) [reply src/gossip/service.zig:838]: created a ticket - https://github.com/orgs/Syndica/projects/2/views/1?pane=issue&itemId=71010101

**Review by 0xNineteen — APPROVED** [review]

**Review by 0xNineteen — APPROVED** [review]

**Pull request merged**
